File: Projects/leetcode/Jump.java
package leetcode;

class Jump {
    public int jump(int[] nums) {
        final int n = nums.length;
        if (n == 1) return 0;
        int currJump = 0;
        int nextMaxReach = 0;
        int currMaxReach = 0;
        for (int i = 0; i < n; i++) {
            nextMaxReach = Math.max(nextMaxReach, i + nums[i]);
            if (nextMaxReach >= n - 1) break;
            if (i == currMaxReach) {
                currJump++;
                currMaxReach = nextMaxReach;
            }
        }
        return currJump + 1;
    }
}

File: Projects/leetcode/jump.kt
package leetcode

fun jump(nums: IntArray): Int {
    val n = nums.size
    if (n == 1) return 0
    var currJump = 0
    var nextMaxReach = 0
    var currMaxReach = 0
    for (i in 0 until n) {
        nextMaxReach = Math.max(nextMaxReach, i + nums[i])
        if (nextMaxReach >= n - 1) break
        if (i == currMaxReach) {
            currJump++
            currMaxReach = nextMaxReach
        }
    }
    return currJump + 1
}
